go/pserver/cclient/cclient.go

```diff
@@ -198,44 +194,58 @@ func paddle_get_params(client C.client, names **C.char, dst **C.paddle_parameter
 		return -1
 	}
 
-	for i := 0; i < int(total); i++ {
-		if i >= len(ps) {
-			break
+	names := func() (string, string) {
+		retNames := ""
+		for _, p := range ps {
+			if retNames == "" {
+				retNames = p.Name
+			} else {
+				retNames = ", " + p.Name
+			}
+		}
+
+		requestedNames := ""
+		for _, n := range ns {
+			if requestedNames == "" {
+				requestedNames = n
+			} else {
+				requestedNames = ", " + n
+			}
+		}
+
+		return requestedNames, retNames
+	}
+
+	if len(ps) != len(ns) {
+		requestedNames, retNames := names()
+		log.Printf("pserver returned wrong number of parameters. Requested: %s, returned: %s.\n", retNames, requestedNames)
+		return -1
+	}
+
+	for i := range ps {
+		if ns[i] != ps[i].Name {
+			requestedNames, retNames := names()
+			log.Printf("pserver returned wrong parameters, or not in requested order. Requested: %s, returned: %s.\n", retNames, requestedNames)
+			return -1
 		}
+	}
 
+	for i := 0; i < int(total); i++ {
 		p := ps[i]
 		param := *(**C.paddle_parameter)(unsafe.Pointer((uintptr(unsafe.Pointer(dst)) + uintptr(i)*unsafe.Sizeof(*dst))))
-		nameReady := false
-		contentAllocated := false
 
 		if unsafe.Pointer(param) == nullPtr {
-			param = (*C.paddle_parameter)(C.calloc(1, C.size_t(unsafe.Sizeof(*param))))
+			log.Println("must pre-allocate parameter.")
+			return -1
```

Comment:
should we define some error code? When I use this client, I have to search the code to make sure the meaning of the ret code